framework/src/tinyalsa/tinyalsa.c

```diff
@@ -1109,53 +1188,134 @@ int pcm_stop(struct pcm *pcm)
 		return -EINVAL;
 	}
 #ifdef CONFIG_AUDIO_MULTI_SESSION
-	if (ioctl(pcm->fd, AUDIOIOC_STOP, (unsigned long)pcm->session) < 0)
+	if (!pcm->running && ioctl(pcm->fd, AUDIOIOC_STOP, (unsigned long)pcm->session) < 0)
```

Comment:
'&&' is correct??? not '||'?
And we don't have to care about pcm->draining?

Reply:
&& is correct. If pcm->running is true, then we will call AUDIOIOC_STOP to stop the codec. Else, we don't call the ioctl since "pcm->running = false" indicates that codec is already stopped. 

We don't consider pcm->draining because:
- in playback scenario, pcm_drain() will be a blocking call. So user cannot call pcm_stop() until the drain operation is completed. 
- pcm->draining variable will be used only for record scenario. In this case, if user calls pcm_stop(), we will just stop the codec and discard the buffers.

Reply:
then I think it is wrong.
you saying that if pcm->running is false, it means codec is already stopped.
Request stop although state is stopped? it is non sense.
It should be requested when pcm->running is true.
But in your code, if it is true, ioctl never called.
You should seperate checking condition.
if (!pcm->running) {
   return some_error or oops
}
if (ioctl) ....
...

Reply:
Oh.. sorry..